File: MATLAB/align_and_validate_dtw_task6_task7.m
function align_and_validate_dtw_task6_task7(est_file, truth_file, output_dir, duration)
%ALIGN_AND_VALIDATE_DTW_TASK6_TASK7  MATLAB equivalent of align_and_validate_dtw_task6_task7.py
%
%   ALIGN_AND_VALIDATE_DTW_TASK6_TASK7(EST_FILE, TRUTH_FILE, OUTPUT_DIR, DURATION)
%   aligns the fused estimator output with ground truth using dynamic time
%   warping, then generates Task 6 overlay plots and basic Task 7 error
%   statistics. The implementation mirrors the Python script of the same
%   name. EST_FILE is a ``*_kf_output.npz`` file. TRUTH_FILE is the
%   corresponding ``STATE_X001.txt``. OUTPUT_DIR defaults to ``results`` and
%   DURATION to 1250 seconds.
%
%   The estimator and truth logs may start at different times and run at
%   different sample rates. ``summarize_timebase`` prints their ranges and
%   effective frequencies so downstream tasks can track these differences.
%
%   This script requires the Signal Processing Toolbox for the DTW
%   implementation and uses ``py.numpy.load`` to read NPZ files.
%
%   The function saves ``aligned_<est_file>.mat`` and ``aligned_<truth_file>``
%   under OUTPUT_DIR and prints summary error statistics.

if nargin < 3 || isempty(output_dir)
    output_dir = get_results_dir();
end
if nargin < 4 || isempty(duration)
    duration = 1250.0;
end
if ~exist(output_dir, 'dir'); mkdir(output_dir); end

% -------------------------------------------------------------------------
% Load estimator data from NPZ
% -------------------------------------------------------------------------
npz = py.numpy.load(string(est_file));
pos_ned = double(npz{'pos_ned_m'});
vel_ned = double(npz{'vel_ned_ms'});
quat = double(npz{'att_quat'});
time_s = double(npz{'time_s'});
ref_lat = double(npz{'ref_lat_rad'});
ref_lon = double(npz{'ref_lon_rad'});
ref_r0 = double(npz{'ref_r0_m'});

% -------------------------------------------------------------------------
% Load truth trajectory
% -------------------------------------------------------------------------
% Truth logs contain a comment header. Columns are:
%   1-count, 2-time [s], 3:5-ECEF position [m], 6:8-ECEF velocity [m/s],
%   9:12-attitude quaternion [q0 q1 q2 q3] (scalar-first)
truth_raw = read_state_file(truth_file);
truth_time = truth_raw(:,2);
truth_pos_ecef = truth_raw(:,3:5);
truth_quat = [];
if size(truth_raw,2) >= 12
    truth_quat = truth_raw(:,9:12);
end

% Convert truth ECEF to NED relative to estimator reference
C_e2n = compute_C_ECEF_to_NED(ref_lat, ref_lon);
truth_pos = (C_e2n * (truth_pos_ecef' - ref_r0)).';

summarize_timebase('Estimator', time_s);
summarize_timebase('Truth', truth_time);

% -------------------------------------------------------------------------
% Smooth and normalise for DTW
% -------------------------------------------------------------------------
pos_est_sm = sgolayfilt(pos_ned, 2, 7);
pos_truth_sm = sgolayfilt(truth_pos, 2, 7);

sig_est = vecnorm(pos_est_sm, 2, 2);
sig_truth = vecnorm(pos_truth_sm, 2, 2);

[~,ix,iy] = dtw(sig_est, sig_truth);

aligned_pos_est = pos_ned(ix,:);
aligned_vel_est = vel_ned(ix,:);
aligned_quat = quat(ix,:);
aligned_pos_truth = truth_pos(iy,:);
common_time = linspace(0, duration, numel(ix)).';

% Save aligned data
[~,ename,~] = fileparts(est_file);
[~,tname,ext] = fileparts(truth_file);
aligned_est_file = fullfile(output_dir, ['aligned_' ename '.mat']);
aligned_truth_file = fullfile(output_dir, ['aligned_' tname ext]);

% Align attitude quaternions if truth is available
aligned_quat_truth = [];
if ~isempty(truth_quat)
    aligned_quat_truth = truth_quat(iy,:);
end

save(aligned_est_file, 'common_time', 'aligned_pos_est', 'aligned_vel_est', ...
    'aligned_quat', 'aligned_quat_truth', 'ref_lat', 'ref_lon', 'ref_r0');
writematrix([common_time aligned_pos_truth], aligned_truth_file);

% -------------------------------------------------------------------------
% Generate plots similar to plot_task6_results.py
% -------------------------------------------------------------------------
plot_task6_results(aligned_pos_est, aligned_vel_est, aligned_quat, common_time, ...
    aligned_pos_truth, common_time, ref_lat, ref_lon, ref_r0, output_dir, ename, aligned_quat_truth);

% If true attitude is available, also plot quaternion error angle over time
if ~isempty(aligned_quat_truth)
    q_est = attitude_tools('quat_hemi', aligned_quat');      % [4 x N]
    q_tru = attitude_tools('quat_hemi', aligned_quat_truth');
    % Ensure both are [N x 4] for dot product computation
    q_est = q_est'; q_tru = q_tru';
    d = sum(q_est .* q_tru, 2);           % dot products
    d = min(1, max(-1, abs(d)));          % hemisphere + clamp
    ang_deg = 2 * acosd(d);
    f = figure('Visible','off','Position',[100 100 700 300]);
    plot(common_time, ang_deg, 'LineWidth', 1.2); grid on;
    xlabel('Time [s]'); ylabel('Attitude Error [deg]');
    title(sprintf('%s Attitude Quaternion Error', ename));
    set(f,'PaperPositionMode','auto');
    pdf_qerr = fullfile(output_dir, sprintf('%s_task6_attitude_quat_error_deg.pdf', ename));
    print(f, pdf_qerr, '-dpdf', '-bestfit'); close(f);
end

% -------------------------------------------------------------------------
% Basic Task 7 error metrics
% -------------------------------------------------------------------------
err = vecnorm(aligned_pos_est - aligned_pos_truth, 2, 2);
fprintf('Aligned time range: 0 to %.2f seconds\n', common_time(end));
fprintf('Position errors: mean=%.4f m, std=%.4f m\n', mean(err), std(err));
end

% -------------------------------------------------------------------------
function plot_task6_results(est_pos, est_vel, est_quat, est_time, ...
    truth_pos, truth_time, ref_lat, ref_lon, ref_r0, out_dir, method, truth_quat)
%PLOT_TASK6_RESULTS  Simplified Task 6 plots in NED frame.
%
%   PLOT_TASK6_RESULTS(EST_POS, EST_VEL, EST_QUAT, EST_TIME, TRUTH_POS,
%   TRUTH_TIME, REF_LAT, REF_LON, REF_R0, OUT_DIR, METHOD) overlays the
%   fused estimator output with the truth data and saves the figure as
%   ``<METHOD>_task6_results_ned.pdf`` under OUT_DIR. METHOD should
%   encode the dataset and algorithm name, e.g.,
%   ``IMU_X001_GNSS_X001_Davenport``. This mirrors ``plot_task6_results.py``.

truth_pos_i = interp1(truth_time, truth_pos, est_time, 'linear', 'extrap');
% Prepare Euler attitude comparison if truth quaternion provided
has_truth_att = (nargin >= 12) && ~isempty(truth_quat);
if has_truth_att
    % Truth quaternion is already aligned to est_time
    eul_est = zeros(numel(est_time),3);
    eul_truth = zeros(numel(est_time),3);
    for k = 1:numel(est_time)
        eul_est(k,:) = quat_to_euler_deg(est_quat(k,:));
        eul_truth(k,:) = quat_to_euler_deg(truth_quat(k,:));
    end
end

labels = {'North','East','Down'};
fig = figure('Visible','off','Position',[100 100 900 800]);
for i=1:3
    subplot(3,3,i); plot(est_time, est_pos(:,i), 'b', est_time, truth_pos_i(:,i), '--r');
    title(['Position ' labels{i}]); ylabel('[m]'); grid on;
    subplot(3,3,3+i); plot(est_time, est_vel(:,i), 'b'); title(['Velocity ' labels{i}]); ylabel('[m/s]'); grid on;
    if has_truth_att
        names = {'Roll','Pitch','Yaw'};
        subplot(3,3,6+i); plot(est_time, eul_est(:,i), 'b', est_time, eul_truth(:,i), '--r');
        title(['Attitude ' names{i}]); ylabel('[deg]'); grid on;
    else
        subplot(3,3,6+i); axis off; text(0.5,0.5,'No truth attitude available','HorizontalAlignment','center');
    end
    if i==3; xlabel('Time [s]'); end
end
sgtitle(sprintf('%s Task 6 Results (NED)', method));
set(fig,'PaperPositionMode','auto');
pdf = fullfile(out_dir, sprintf('%s_task6_results_ned.pdf', method));
print(fig, pdf, '-dpdf');
close(fig);
end

function C = compute_C_ECEF_to_NED(lat_rad, lon_rad)
%COMPUTE_C_ECEF_TO_NED  Local rotation matrix from ECEF to NED.
s_lat = sin(lat_rad); c_lat = cos(lat_rad);
s_lon = sin(lon_rad); c_lon = cos(lon_rad);
C = [-s_lat*c_lon, -s_lat*s_lon,  c_lat;
     -s_lon,       c_lon,       0;
     -c_lat*c_lon, -c_lat*s_lon, -s_lat];
end

function summarize_timebase(label, t)
%SUMMARIZE_TIMEBASE  Print basic timing statistics for a dataset.
if numel(t) < 2
    fprintf('%s time vector too short for statistics\n', label);
    return;
end

function eul_deg = quat_to_euler_deg(q)
%QUAT_TO_EULER_DEG Convert [q0 q1 q2 q3] to XYZ Euler [deg].
    q0=q(1); q1=q(2); q2=q(3); q3=q(4);
    R=[1-2*(q2^2+q3^2) 2*(q1*q2-q0*q3) 2*(q1*q3+q0*q2);...
       2*(q1*q2+q0*q3) 1-2*(q1^2+q3^2) 2*(q2*q3-q0*q1);...
       2*(q1*q3-q0*q2) 2*(q2*q3+q0*q1) 1-2*(q1^2+q2^2)];
    roll  = atan2(R(3,2), R(3,3));
    pitch = -asin(R(3,1));
    yaw   = atan2(R(2,1), R(1,1));
    eul_deg = rad2deg([roll, pitch, yaw]);
end
dt = diff(t);
mean_dt = mean(dt);
freq = 1./mean_dt;
fprintf('%s time start=%.3f s end=%.3f s mean_dt=%.3f s (%.2f Hz)\n', ...
    label, t(1), t(end), mean_dt, freq);
end

File: python/src/align_and_validate_dtw_task6_task7.py
# Updated script for aligning IMU/GNSS estimate with truth using DTW
# and generating Task 6/7 plots. Designed for IMU_X001_GNSS_X001_TRIAD
# filter output. Estimator and truth logs may have different time bases;
# ``summarize_timebase`` prints their ranges and effective sample rates.

import argparse
import os
import subprocess

import matplotlib.pyplot as plt
import numpy as np
from fastdtw import fastdtw
from scipy.interpolate import interp1d
from scipy.signal import savgol_filter
from scipy.spatial.distance import euclidean


# ---------------------------------------------------------------------------
# Utility functions
# ---------------------------------------------------------------------------

def smooth_data(data: np.ndarray, window_length: int = 7, polyorder: int = 2) -> np.ndarray:
    """Apply Savitzky-Golay filter to reduce noise."""
    if len(data) < window_length:
        return data
    return savgol_filter(data, window_length, polyorder, axis=0)


def load_estimator_data(path: str):
    """Return position, velocity, quaternion, time and reference parameters."""
    data = np.load(path, allow_pickle=True)
    try:
        pos = data["pos_ned_m"]
        vel = data["vel_ned_ms"]
        quat = data["att_quat"]
        time = data["time_s"]
        ref_lat = data["ref_lat_rad"]
        ref_lon = data["ref_lon_rad"]
        ref_r0 = data["ref_r0_m"]
    except KeyError as exc:  # pragma: no cover - debugging helper
        print(f"Missing key {exc} in {path}. Available keys: {list(data.keys())}")
        raise
    print("Using 'pos_ned_m' for position, 'vel_ned_ms' for velocity, 'att_quat' for quaternion")
    return pos, vel, quat, time, ref_lat, ref_lon, ref_r0


def load_truth_data(path: str):
    """Return position and time from ``STATE_X*.txt`` truth logs.

    The text files start with a sample counter followed by the timestamp and
    ECEF coordinates.  Earlier versions of this helper mistakenly treated the
    counter as the time column, leading to misaligned overlays when the first
    column was used.  We now explicitly skip the counter and extract the time
    from column 1 and the position from columns 2–4.
    """
    data = np.loadtxt(path)
    time = data[:, 1]
    pos = data[:, 2:5]
    frame = "ned" if np.abs(pos).max() < 1e6 else "ecef"
    return pos, time, frame


def summarize_timebase(label: str, time: np.ndarray) -> None:
    """Print basic timing statistics for a dataset.

    Parameters
    ----------
    label : str
        Identifier for the dataset (e.g. ``"Estimator"``).
    time : np.ndarray
        Time vector in seconds.
    """
    if time.size < 2:
        print(f"{label} time vector too short for statistics")
        return
    dt = np.diff(time)
    mean_dt = np.mean(dt)
    freq = 1.0 / mean_dt if mean_dt > 0 else float('inf')
    print(
        f"{label} time start={time[0]:.3f}s end={time[-1]:.3f}s "
        f"mean_dt={mean_dt:.3f}s ({freq:.2f} Hz)"
    )


def ned_to_ecef(pos_ned: np.ndarray, lat: float, lon: float, r0: np.ndarray) -> np.ndarray:
    """Convert NED positions to ECEF using reference lat/lon/r0."""
    C = np.array(
        [
            [-np.sin(lat) * np.cos(lon), -np.sin(lon), -np.cos(lat) * np.cos(lon)],
            [-np.sin(lat) * np.sin(lon), np.cos(lon), -np.cos(lat) * np.sin(lon)],
            [np.cos(lat), 0.0, -np.sin(lat)],
        ]
    )
    return (r0 + C.T @ pos_ned.T).T


def quat_to_dcm(quat: np.ndarray) -> np.ndarray:
    """Convert quaternion [w,x,y,z] to a DCM."""
    w, x, y, z = quat
    return np.array(
        [
            [1 - 2 * (y ** 2 + z ** 2), 2 * (x * y - w * z), 2 * (x * z + w * y)],
            [2 * (x * y + w * z), 1 - 2 * (x ** 2 + z ** 2), 2 * (y * z - w * x)],
            [2 * (x * z - w * y), 2 * (y * z + w * x), 1 - 2 * (x ** 2 + y ** 2)],
        ]
    )


# ---------------------------------------------------------------------------
# Plot generation
# ---------------------------------------------------------------------------

def plot_task6_results(
    est_pos: np.ndarray,
    est_vel: np.ndarray,
    est_quat: np.ndarray,
    est_time: np.ndarray,
    truth_pos: np.ndarray,
    truth_time: np.ndarray,
    ref_lat: float,
    ref_lon: float,
    ref_r0: np.ndarray,
    output_dir: str,
    method_name: str,
) -> None:
    """Create and save Task 6 plots matching Task 5 layout."""

    truth_interp = interp1d(truth_time, truth_pos, axis=0, bounds_error=False, fill_value="extrapolate")
    truth_pos_i = truth_interp(est_time)

    dt = est_time[1] - est_time[0] if len(est_time) > 1 else 1.0
    est_acc = np.diff(est_vel, axis=0) / dt
    est_acc = np.vstack([est_acc, est_acc[-1]])

    # Plot 1 - NED position/velocity/acceleration
    plt.figure(figsize=(15, 10))
    labels = ["North", "East", "Down"]
    for i, lab in enumerate(labels):
        plt.subplot(3, 3, i + 1)
        plt.plot(est_time, est_pos[:, i], label="Fused")
        plt.plot(est_time, truth_pos_i[:, i], "--", label="Truth")
        plt.xlabel("Time (s)")
        plt.ylabel(f"Position {lab} (m)")
        plt.legend(); plt.grid(True)
    for i, lab in enumerate(labels):
        plt.subplot(3, 3, i + 4)
        plt.plot(est_time, est_vel[:, i], label="Fused")
        plt.xlabel("Time (s)"); plt.ylabel(f"Velocity {lab} (m/s)")
        plt.legend(); plt.grid(True)
    for i, lab in enumerate(labels):
        plt.subplot(3, 3, i + 7)
        plt.plot(est_time, est_acc[:, i], label="Fused")
        plt.xlabel("Time (s)"); plt.ylabel(f"Acceleration {lab} (m/s²)")
        plt.legend(); plt.grid(True)
    plt.tight_layout()
    out_path = os.path.join(output_dir, f"{method_name}_task6_results_ned.pdf")
    plt.savefig(out_path)
    plt.close()

    # Plot 2 - NED position & body-frame velocity
    vel_body = np.zeros_like(est_vel)
    for i, q in enumerate(est_quat):
        vel_body[i] = quat_to_dcm(q).T @ est_vel[i]
    plt.figure(figsize=(15, 10))
    for i, lab in enumerate(labels):
        plt.subplot(2, 3, i + 1)
        plt.plot(est_time, est_pos[:, i], label="Fused")
        plt.plot(est_time, truth_pos_i[:, i], "--", label="Truth")
        plt.xlabel("Time (s)"); plt.ylabel(f"Position {lab} (m)")
        plt.legend(); plt.grid(True)
    body_labels = ["X", "Y", "Z"]
    for i, lab in enumerate(body_labels):
        plt.subplot(2, 3, i + 4)
        plt.plot(est_time, vel_body[:, i], label="Fused")
        plt.xlabel("Time (s)"); plt.ylabel(f"Velocity {lab} (m/s)")
        plt.legend(); plt.grid(True)
    plt.tight_layout()
    out_path = os.path.join(output_dir, f"{method_name}_task6_results_mixed.pdf")
    plt.savefig(out_path)
    plt.close()

    # Plot 3 - All data in NED frame
    plt.figure(figsize=(15, 5))
    for i, lab in enumerate(labels):
        plt.subplot(1, 3, i + 1)
        plt.plot(est_time, est_pos[:, i], label="Fused")
        plt.plot(est_time, truth_pos_i[:, i], "--", label="Truth")
        plt.xlabel("Time (s)"); plt.ylabel(f"Position {lab} (m)")
        plt.legend(); plt.grid(True)
    plt.tight_layout()
    out_path = os.path.join(output_dir, f"{method_name}_task6_results_all_ned.pdf")
    plt.savefig(out_path)
    plt.close()

    # Plot 4 - All data in ECEF frame
    est_ecef = ned_to_ecef(est_pos, ref_lat, ref_lon, ref_r0)
    truth_ecef = ned_to_ecef(truth_pos_i, ref_lat, ref_lon, ref_r0)
    plt.figure(figsize=(15, 5))
    ecef_labels = ["X", "Y", "Z"]
    for i, lab in enumerate(ecef_labels):
        plt.subplot(1, 3, i + 1)
        plt.plot(est_time, est_ecef[:, i], label="Fused")
        plt.plot(est_time, truth_ecef[:, i], "--", label="Truth")
        plt.xlabel("Time (s)"); plt.ylabel(f"Position {lab} (m)")
        plt.legend(); plt.grid(True)
    plt.tight_layout()
    out_path = os.path.join(output_dir, f"{method_name}_task6_results_all_ecef.pdf")
    plt.savefig(out_path)
    plt.close()

    # Plot 5 - All data in body frame
    pos_body = np.zeros_like(est_pos)
    for i, q in enumerate(est_quat):
        pos_body[i] = quat_to_dcm(q).T @ est_pos[i]
    plt.figure(figsize=(15, 5))
    for i, lab in enumerate(body_labels):
        plt.subplot(1, 3, i + 1)
        plt.plot(est_time, pos_body[:, i], label="Fused")
        plt.xlabel("Time (s)"); plt.ylabel(f"Position {lab} (m)")
        plt.legend(); plt.grid(True)
    plt.tight_layout()
    out_path = os.path.join(output_dir, f"{method_name}_task6_results_all_body.pdf")
    plt.savefig(out_path)
    plt.close()


# ---------------------------------------------------------------------------
# Main entry point
# ---------------------------------------------------------------------------

def main() -> None:
    ap = argparse.ArgumentParser(description="Align and validate datasets using DTW")
    ap.add_argument("--est-file", required=True, help="Path to *_kf_output.npz file")
    ap.add_argument("--truth-file", required=True, help="Path to STATE_X001.txt file")
    ap.add_argument("--output-dir", default="results", help="Directory for outputs")
    ap.add_argument("--duration", type=float, default=1250.0, help="Expected duration")
    args = ap.parse_args()

    os.makedirs(args.output_dir, exist_ok=True)

    est_pos, est_vel, est_quat, est_time, ref_lat, ref_lon, ref_r0 = load_estimator_data(args.est_file)
    truth_pos, truth_time, truth_frame = load_truth_data(args.truth_file)

    summarize_timebase("Estimator", est_time)
    summarize_timebase("Truth", truth_time)

    if truth_frame != "ned":
        raise RuntimeError("Truth trajectory must be in NED frame")

    # Smooth for alignment
    est_pos = smooth_data(est_pos)
    truth_pos = smooth_data(truth_pos)

    eps = 1e-10
    est_norm = (est_pos - np.median(est_pos, axis=0)) / (np.std(est_pos, axis=0) + eps)
    truth_norm = (truth_pos - np.median(truth_pos, axis=0)) / (np.std(truth_pos, axis=0) + eps)

    print("Starting Task 6 overlay ...")
    distance, path = fastdtw(est_norm, truth_norm, dist=euclidean)
    est_idx, truth_idx = zip(*path)

    n = min(len(est_idx), len(truth_idx), len(est_time))
    common_time = np.linspace(0.0, args.duration, n)
    est_interp = interp1d(np.linspace(0.0, args.duration, len(est_idx)), est_pos[list(est_idx)], axis=0)
    truth_interp = interp1d(np.linspace(0.0, args.duration, len(truth_idx)), truth_pos[list(truth_idx)], axis=0)
    est_vel_i = interp1d(np.linspace(0.0, args.duration, len(est_idx)), est_vel[list(est_idx)], axis=0)
    est_quat_i = interp1d(np.linspace(0.0, args.duration, len(est_idx)), est_quat[list(est_idx)], axis=0)

    aligned_est_pos = est_interp(common_time)
    aligned_truth_pos = truth_interp(common_time)
    aligned_est_vel = est_vel_i(common_time)
    aligned_est_quat = est_quat_i(common_time)

    aligned_est_file = os.path.join(args.output_dir, "aligned_" + os.path.basename(args.est_file))
    np.savez(
        aligned_est_file,
        time_s=common_time,
        pos_ned_m=aligned_est_pos,
        vel_ned_ms=aligned_est_vel,
        att_quat=aligned_est_quat,
        ref_lat_rad=ref_lat,
        ref_lon_rad=ref_lon,
        ref_r0_m=ref_r0,
    )

    aligned_truth_file = os.path.join(args.output_dir, "aligned_" + os.path.basename(args.truth_file))
    np.savetxt(
        aligned_truth_file,
        np.column_stack((common_time, aligned_truth_pos)),
        fmt="%.6f",
        header="time x y z",
    )

    method_name = os.path.basename(args.est_file).replace("_kf_output.npz", "")
    plot_task6_results(
        aligned_est_pos,
        aligned_est_vel,
        aligned_est_quat,
        common_time,
        truth_pos,
        truth_time,
        ref_lat,
        ref_lon,
        ref_r0,
        args.output_dir,
        method_name,
    )

    print("Running Task 7 evaluation ...")
    cmd = [
        "python",
        "src/validate_3sigma.py",
        "--est-file",
        aligned_est_file,
        "--truth-file",
        aligned_truth_file,
        "--output-dir",
        args.output_dir,
    ]
    res = subprocess.run(cmd, capture_output=True, text=True)
    print("Validation Output:")
    print(res.stdout)
    print(res.stderr)

    pos_err = np.linalg.norm(aligned_est_pos - aligned_truth_pos, axis=1)
    print(f"Aligned time range: 0 to {common_time[-1]:.2f} seconds")
    print(f"Position errors: mean={np.mean(pos_err):.4f} m, std={np.std(pos_err):.4f} m")


if __name__ == "__main__":
    main()
